apheleia.el

```diff
@@ -289,7 +319,109 @@ at once.")
+      (process-send-string
+       proc
+       (with-current-buffer stdin
+         (buffer-string))))
+    (process-send-eof proc)
+    proc))
+
+(cl-defun apheleia--call-process
+    (&key name stdin stdout stderr command remote noquery callback)
+  "Helper to synchronously run a formatter process.
+This function essentially runs COMMAND synchronously passing STDIN
+as standard input and saving output to the STDOUT and STDERR buffers.
+Once the process is finished CALLBACK will be invoked with the exit
+code (see `process-exit-status') of the process.
+
+This function accepts all the same arguments as `apheleia--make-process'
+for simplicity, however some may not be used. This includes: NAME, and
+NO-QUERY."
+  (ignore name noquery)
+  (let* ((stderr-file (apheleia--make-temp-file nil "apheleia"))
+         (args
+          (list (car command)            ; argv[0]
+                (not stdin)              ; If stdin we don't delete the STDIN buffer text
+                                         ; with `call-process-region'. Otherwise we send
+                                         ; no INFILE argument to `call-process'.
+                `(,stdout ,stderr-file)  ; stdout buffer and stderr file. `call-process'
+                                         ; cannot capture stderr into a separate buffer,
+                                         ; the best we can do is save and read from a file.
+                nil                      ; Do not re/display stdout as output is recieved
+                (cdr command)            ; argv[1:]
+                )))
+    (unwind-protect
+        (let ((exit-status
+               (cond
+                ((and remote stdin)
+                 ;; There's no call-process variant for this, we'll have to
+                 ;; copy STDIN to a remote temporary file, create a subshell
+                 ;; on the remote that runs the formatter and passes the temp
+                 ;; file as stdin and then deletes it.
+                 (let* ((remote-stdin
+                         (apheleia--make-temp-file remote "apheleia-stdin"))
+                        ;; WARN: This assumes a POSIX compatible shell.
+                        (shell (or (bound-and-true-p tramp-default-remote-shell)
+                                   "sh"))
+                        (shell-command
+                         (concat
+                          (mapconcat #'shell-quote-argument command " ")
+                          " < "
+                          (shell-quote-argument
+                           (apheleia--strip-remote remote-stdin)))))
+                   (unwind-protect
+                       (progn
+                         (with-current-buffer stdin
+                           (apheleia--write-region-silently nil nil
+                                                            remote-stdin))
+
+                         (process-file
+                          shell nil (nth 2 args) nil "-c" shell-command))
```

Comment:
Copying happens here, black is saying `error: cannot format -: '\x1b'` and the remote file ends with a `^[[0m` for some reason/

Reply:
Weird -- one thing I'll note is `\x1b[0m` is the ANSI control sequence for "reset color" -- maybe TRAMP adds that for some reason because it's trying to reuse the same shell session and doesn't want the color to be messed up by a poorly behaved program?